File: 189_rotate_array/extra_copy_array.py
class Solution:
    def rotate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        """
        Complexity: O(n)
        Space: O(n)
        Created one extra copy of original list, and used it for the assignment 
        of the rotating list.
        """
        ori_nums = nums[:]
        
        # need to mod len in case the k is biggen than len(nums)
        for idx in range(len(nums)):
            nums[idx] = ori_nums[idx-(k % len(nums))]

File: 189_rotate_array/extra_copy_array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        vector<int>copied(nums);
        for(int i = 0; i<nums.size(); ++i){
            int idx = (i < (k % nums.size()) ? nums.size()- (k % nums.size() - i) : i - (k % nums.size()));
            nums[i] = copied[idx];
        }
    }
};
